### PR TITLE
replace gorm with dal from services module

### PR DESCRIPTION
**Pull request by klesh**
### Summary

replace `gorm` with `dal` from `services` module

### Does this close any open issues?
prerequisite of #3709 

## Breaking Changes

1. rerun task: `POST /tasks/:taskId/rerun`
2. rerun pipeline failed tasks: `POST /pipelines/:pipelineId/rerun`

### pipeline

- create/cancel pipeline, the following actions were taken and verified
  - PIPELINE_MAX_PARALLEL="1" create 2 pipelines by using `scripts/pm/github/trigger-pipline.sh`: **succeeded**
  - cancel the pipeline with "status=CREATED": **succeeded**
  - cancel the running pipeline: **succeeded**
  - PIPELINE_MAX_PARALLEL="2": #3978 (not related to this PR)
  - check progress field for pipeline/tasks: **succeeded**
  - create pipeline with label=foobar
- list pipelines, the following urls were requested and **verified**
  - "pipelines" : **succeeded**
  - "pipelines?pageSize=1&page=1" : **succeeded**
  - "pipelines?pageSize=1&page=2" : **succeeded**
  - "pipelines?blueprint_id=1": **succeeded**
  - "pipelines?status=TASK_FAILED": **succeeded**
  - "pipelines?pending=1": **succeeded**
  - "pipelines?label=foobar": **succeeded**
- rerun:
  - test rerun single task: **succeeded**
  - test rerun pipeline: **succeeded**

### blueprints

- create/trigger blueprints, the following actions were taken and verified
  - create blueprint without `project_name`: **succeeded**
  - trigger the blueprint: **succeeded**
- list blueprints, the following urls were requested and **verified**
  - "blueprints"
  - "blueprints?pageSize=1&page=1" 
  - "blueprints?pageSize=1&page=2"
  - "blueprints?enable=1"
  - "blueprints?is_manual=0"
  - "blueprints?label=foobar"

### project

- create new project
  - create new project with dora: **succeeded**
  - check the `pipeline.plan` with dora: **succeeded**
  - disable dora for the project: **succeeded**
  - check the `pipeline.plan` without dora: **succeeded**

### CODE REVIEW

File: impl/dalgorm/dalgorm.go [hunk 1]
```diff
@@ -360,6 +360,11 @@ func (d *Dalgorm) IsErrorNotFound(err errors.Error) bool {
 	return errors.Is(err, gorm.ErrRecordNotFound)
 }
 
+// IsDuplicationError checking if the sql error is not found.
+func (d *Dalgorm) IsDuplicationError(err errors.Error) bool {
+	return errors.Is(err, gorm.ErrRecordNotFound)
```
Comment (mindlesscloud) [inline impl/dalgorm/dalgorm.go:365]: Why the `gorm.ErrRecordNotFound` is duplication error?

File: services/blueprint_makeplan_v200.go [hunk 2]
```diff
@@ -41,11 +42,17 @@ func GeneratePlanJsonV200(
 	}
 	// refresh project_mapping table to reflect project/scopes relationship
 	if len(scopes) > 0 {
+		if projectName != "" {
```
Comment (likyh) [inline services/blueprint_makeplan_v200.go:45]: rebase error. it have been moved to Line#59

File: services/dm_code.go [hunk 3]
```diff
@@ -20,20 +20,12 @@ package services
...
-	if err != nil {
-		return nil, 0, errors.Convert(err)
-	}
-	err = db.Find(&repos).Error
-	if err != nil {
-		return nil, count, errors.Convert(err)
-	}
-	return repos, count, nil
+	err := db.All(&repos, dal.Orderby("id DESC"))
+	return repos, int64(len(repos)), err
```
Comment (likyh) [inline services/dm_code.go:30]: only 1 page?

File: services/project.go [hunk 4]
```diff
@@ -22,286 +22,275 @@ import (
...
 	if err != nil {
 		return nil, err
 	}
 
-	err = helper.DecodeMapStruct(body, projectInput)
+	// wrap all operation inside a transaction
+	tx := db.Begin()
+	defer func() {
+		if r := recover(); r != nil || err != nil {
+			err = tx.Rollback()
```
Comment (likyh) [inline services/project.go:146]: it seems no use because there is no panic.

File: services/dm_code.go (hunk 3 above)
Comment (klesh) [reply services/dm_code.go:30]: not really sure what this API does, it was the original logic

File: services/project.go (hunk 4 above)
Comment (klesh) [reply services/project.go:146]: no sure when and why would this happen. maybe the db went off? lets ignore it at the time being

Comment (likyh) [reply services/project.go:146]: And rollback may not be called if other err returned?

**Review by likyh — APPROVED** [review]

**Pull request merged**